nbodykit/source/catalog/tests/test_hod.py

```diff
@@ -1,175 +1,88 @@
-    source = LogNormalCatalog(Plin=Plin, nbar=3e-3, BoxSize=BoxSize, Nmesh=128, seed=42)
-
-    # run FOF
-    r = FOF(source, linking_length=0.2, nmin=20)
-    halos = r.to_halos(cosmo=cosmo, redshift=redshift, particle_mass=1e12, mdef='vir', posdef='cm')
+    # the demo halos
+    halos = DemoHaloCatalog('bolshoi', 'rockstar', 0.5)
 
-    # add an object column to the input halo catalog
-    cat = halos.to_halotools()
-    test =  numpy.random.random(size=len(cat.halo_table)).astype('O')
-    cat.halo_table['test'] = test
+    # initialize model with bad MASS column
+    model = Zheng07Model.to_halotools(halos.cosmo, halos.attrs['redshift'], '200c')
 
-    # make the HOD
-    with pytest.raises(TypeError):
-        hod = HODCatalog(cat, seed=42)
+    with pytest.raises(Exception):
+        hod = halos.populate(model)
 
 
 @MPITest([4])
-def test_bad_catalog(comm):
+def test_no_galaxies(comm):
 
     CurrentMPIComm.set(comm)
+    halos = DemoHaloCatalog('bolshoi', 'rockstar', 0.5)
 
-    redshift = 0.55
-    cosmo = cosmology.Planck15
-    BoxSize = 512
-
-    Plin = cosmology.LinearPower(cosmo, redshift=redshift, transfer='EisensteinHu')
-    source = LogNormalCatalog(Plin=Plin, nbar=3e-3, BoxSize=BoxSize, Nmesh=128, seed=42)
-
-    # run FOF
-    r = FOF(source, linking_length=0.2, nmin=20)
-    halos = r.to_halos(cosmo=cosmo, redshift=redshift, particle_mass=1e12, mdef='vir', posdef='cm')
-
-    # test converting from astropy to nbodykit cosmology
-    cat = halos.to_halotools()
-    cat.cosmology = cat.cosmology.to_astropy()
-
-    # make the HOD
-    hod = HODCatalog(cat, seed=42)
-
-    # missing required column
-    del cat.halo_table['halo_id']
+    # no galaxies populated into halos
     with pytest.raises(ValueError):
-        hod = HODCatalog(cat, seed=42)
-
-    # missing required attribute
-    del cat.cosmology
-    with pytest.raises(AttributeError):
-        hod = HODCatalog(cat, seed=42)
-
+        hod = halos.populate(Zheng07Model, seed=42, logMmin=16)
 
 @MPITest([4])
 def test_repopulate(comm):
 
     CurrentMPIComm.set(comm)
+    halos = DemoHaloCatalog('bolshoi', 'rockstar', 0.5)
 
-    redshift = 0.55
-    cosmo = cosmology.Planck15
-    BoxSize = 512
-
-    Plin = cosmology.LinearPower(cosmo, redshift=redshift, transfer='EisensteinHu')
-    source = LogNormalCatalog(Plin=Plin, nbar=3e-3, BoxSize=BoxSize, Nmesh=128, seed=42)
-
-    # run FOF
-    r = FOF(source, linking_length=0.2, nmin=20)
-    halos = r.to_halos(cosmo=cosmo, redshift=redshift, particle_mass=1e12, mdef='vir', posdef='cm')
+    # do not call repopulate first
+    with pytest.raises(ValueError):
+        hod = halos.repopulate()
 
-    # make the HOD catalog from halotools catalog
-    hod = HODCatalog(halos.to_halotools(), seed=42)
+    # populate the mock first
+    hod = halos.populate(Zheng07Model, seed=42)
     size = hod.csize
 
     # repopulate (with same seed --> same catalog)
-    hod.repopulate(seed=42)
+    hod = halos.repopulate(seed=42)
```

Comment:
do you like hod.repopulate() better than halos.repopulate()?

Reply:
If that's the case then the first ValueError can be eliminated -- without doing populate first there is no hod object for you to repopulate. 

The implication is that the compiled model shall be associated with hod, not halos.